Installation from source review

stevepiercy opened the pull request:
This PR is both a review, builds upon, and replaces https://github.com/plone/documentation/pull/1274

ksuess committed:
Apply Steve's review comments on 'installation from packages'

stevepiercy committed:
Fix linkcheck issues

stevepiercy committed:
Tidy up Glossary

stevepiercy committed:
Clean up writing-docs-guide.md
- Remove toggle per https://github.com/plone/documentation/pull/1260#discussion_r909494893. This also simplifies the nested directive example.
-

stevepiercy committed:
English grammar

stevepiercy committed:
Overhaul Install index to not suck so hard
Add sphinx-design to requirements
Add sphinx_design to conf.py
Tidy up docs referenced by Install Index
Add Extensions section to writing-docs-guide.md

stevepiercy committed:
Break paragraphs apart into logical parts and list.

stevepiercy committed:
Convert item to a TODO directive

stevepiercy committed:
Merge branch '6-dev' into installation-from-source-review

# Conflicts:
#	docs/glossary.md
#	docs/install/installation-backend-from-packages-step-by-step.md

stevepiercy committed:
Merge branch '6-dev' into installation-from-source-review

stevepiercy committed:
Commit WIP

stevepiercy committed:
Rewrite "Install Plone from its packages" to be streamlined.
Remove duplicated or obsolete content, replacing duplicated content with links to source material.
Reorganize the remaining bits into a new supplementary page "Manage add-ons, packages, and processes".
Add Make and cookiecutter-plone-starter to the Glossary.
Remove now unused toggle style and JavaScript.
Optionally style images similar to figures, with <img class="figure" ... >

stevepiercy committed:
Complete work on moving content from install-backend-from-packages-step-by-step.md to manage-add-ons-packages.md

stevepiercy committed:
Merge branch '6-dev' into installation-from-source-review

# Conflicts:
#	docs/install/source.md

stevepiercy committed:
Ignore redirect that requires login to GitHub

stevepiercy committed:
Fix redirect of Classic UI demo site

stevepiercy commented:
Two other issues also need to be reviewed and merged in other repos.

- https://github.com/plone/volto/pull/3660
- https://github.com/plone/plone.restapi/pull/1495

jensens reviewed (dismissed):
It is more than installation from source which is touched, right? But anyway, I would say go for merge,

ksuess commented on docs/install/install-from-packages.md:
````diff
@@ -0,0 +1,291 @@
+cd my_project
+```
+
+Issue the following commands in a shell session to create a Python virtual environment in the current directory.
+
+```shell
+python -m venv venv
+source venv/bin/activate
+```
+
+Upgrade Python package management tools.
+
+```shell
+pip install --upgrade pip wheel
+```
````
No need for a Python virtual environment, this is done by cookiecutter-plone-starter.

stevepiercy replied:
Yes, there are two virtual environments. This one is for installing cookiecutter, which in turn creates the second.

Instead should we direct users to install cookiecutter in their user Python? That's what I normally do.

Options:

```
pip install --user --upgrade cookiecutter
python3 -m pip install --user --upgrade cookiecutter
python3 -m pip install --user --upgrade cookiecutter pip wheel
```

jensens replied:
Resently I tend to teach using pipx. It really simplify usage of this tool category. 
What about a global pipx install and then use "pipx run cookiecutter". Same is valid for things like black, isort, name it.

ale-rt replied:
I am also a big fan of [pipx](https://github.com/pypa/pipx).
It is also true that `pip install --user` is more "universal".

stevepiercy replied:
I don't have an opinion either way. I just want everyone to agree on "the one way" to do things. hahaha! I know, silly me.

I would say let's hold off on adopting `pipx`. We are focusing on getting Plone 6 docs out the door (holy cow, PloneConf is in 2 weeks and we still don't have a 6 release 😱 ). Taking time to shift to `pipx` from `pip` everywhere would delay that. I think we should revisit and discuss the use of packaging tools at PloneConf. I would be happy to accept PRs to shift to whatever is agreed upon at a docs sprint.

ale-rt replied:
+1
I think that pipx user can easily figure out how to adjust the instructions to use it.

stevepiercy replied:
Fixed in e27ace149791a496a1120c3eaaa2e614ea2dfbc3

ksuess commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,511 @@
+-   Creates specified directories, if they do not exist.
+
+As you may surmise, you can configure your Zope instance.
+You have two options for configuration.
```
Developer does not need to make a decision on one option. `cookecutter-plone-starter`already generated /instance/etc/ with configuration files for Zope. These configuration files can be customized now.

stevepiercy replied:
So... would deleting everything between:

```
-   Creates specified directories, if they do not exist.
```

...and...

```
(manage-zope-configuration-with-cookiecutter-zope-instance-label)=
```

and maybe inserting this:

```
You can configure your Zope instance using `make` as described in the next section.
```

...be the way to go?

stevepiercy replied:
Fixed in 315dea1c08d0b75a2b52e6b4fd485f2f7005c1bb

ksuess commented on docs/install/manage-add-ons-packages.md:
````diff
@@ -0,0 +1,511 @@
+
+Then after specifying requirements, you can install them by changing your working directory to `backend`, and using the following command.
+
+```shell
+cd backend
+bin/pip install -r requirements.txt
+```
+
+Next you can use `cookiecutter` to generate the Zope configuration files.
+
+```shell
+bin/cookiecutter -f --no-input --config-file instance.yaml https://github.com/plone/cookiecutter-zope-instance
+```
+
+Let's break down that command.
+
+```{list-table} Table of Cookiecutter arguments
+:header-rows: 1
+
+* - Option
+  - Effect
+* - `-f`, `--overwrite-if-exists`
+  - Overwrite the contents of the output directory if it already exists.
+    Data in `var` stays untouched.
+* - `--no-input`
+  - Do not prompt for parameters and only use `cookiecutter.json` file content.
+* - `--config-file instance.yaml`
+  - User configuration file.
+```
+
+As we mentioned earlier in {ref}`manage-installation-details-label`, `cookiecutter-zope-instance` creates a directory {file}`instance` that contains your Zope and Plone instance configuration.
+This directory also contains persistent storage objects. During {doc}`install-from-packages`, when we ran the command `make start-backend`, the `Makefile` target ran the following command.
+
+```shell
+runwsgi -v instance/etc/zope.ini
+```
+
+That command loaded that instance's configuration, and started the Zope instance and our Plone site.
+However, if your site is still running in the backend shell session, the changes will not show up until you restart the Plone backend.
+If needed, stop your Plone backend with {kbd}`ctrl-c`.
````
Future branch of 
cookiecutter-plone.starter branch https://github.com/collective/cookiecutter-plone-starter/tree/feature-4 will provide `make build-backend` and in short `make build` to update build after changes (like additional add-on)

Replace with
`make build`

stevepiercy replied:
Yeah, I added a `make upgrade` target to do the same thing. `make build` currently tries to create a site, and it errored out. I like the new version much better. I'll run through it tonight and try it out.

Notes for me:

https://github.com/collective/cookiecutter-plone-starter/blob/2c269e73ef869389bfd5ac7e4b9296efa6b67c5d/%7B%7B%20cookiecutter.project_slug%20%7D%7D/backend/Makefile#L76-L81

Specify branch when invoking cookiecutter.

stevepiercy replied:
Fixed in e27ace149791a496a1120c3eaaa2e614ea2dfbc3

ksuess commented on docs/install/manage-add-ons-packages.md:
````diff
@@ -0,0 +1,511 @@
+
+{file}`mx.ini`
+
+```ini
+[settings]
+# constraints of Plone packages
+version-overrides =
+    plone.api>=2.0.0a3
+
+[plone.restapi]
+url = git@github.com:plone/plone.restapi.git
+branch = master
+extras = test
+```
+
+With these three files in your project, install and run `mxdev` with the following commands.
+
+```shell
+pip install mxdev
+mxdev -c mx.ini
+```
+
+`mxdev` generates the files {file}`requirements-mxdev.txt` and {file}`constraints-mxdev.txt`.
+Now you can install your packages with `pip` and the new requirements file:
+
+```shell
+pip install -r requirements-mxdev.txt
+```
+
+Finally, to reload the packages, restart your Zope instance/Plone site with the following command.
+
+```shell
+runwsgi instance/etc/zope.ini
+```
+
+```{seealso}
+This was a brief overview of how `mxdev` helps with versions and checkouts.
+It can do a lot more.
+See the [documentation of `mxdev` in its README.rst](https://github.com/mxstack/mxdev/blob/main/README.rst) for complete information.
+```
````
Comment: Future branch of
cookiecutter-plone.starter branch https://github.com/collective/cookiecutter-plone-starter/tree/feature-4 will provide mxdev and it's configuration files like described above. 

The configuration files just need to be customized.

`make build` or `make build-backend` builds after changes.
`make start` or `make start-backen`starts rebuild instance.

stevepiercy replied:
Fixed in 6c9128b1d5038f723e0b8af65fcbc2b0b9a571f0

ksuess commented:
This is importent: https://deploy-preview-1277--6-dev-docs-plone-org.netlify.app/install/manage-add-ons-packages.html#common-management-tasks. With further changes, I think this section should be very well present.

ksuess commented:
Please be aware that this documentation section depends on the decision that `cookiecutter-plone-starter`is the way to move forward for an 'all-in-one' (backend and frontend) installer of Plone 6 from packages.

ksuess commented:
> Please be aware that this documentation section depends on the decision that `cookiecutter-plone-starter`is the way to move forward for an 'all-in-one' (backend and frontend) installer of Plone 6 from packages.

+1

stevepiercy commented:
> Please be aware that this documentation section depends on the decision that `cookiecutter-plone-starter`is the way to move forward for an 'all-in-one' (backend and frontend) installer of Plone 6 from packages.

Indeed. I rarely get excited about software, but this way of doing things was so compelling that I tried to explain it to my wife over dinner last night.

ale-rt commented:
> > Please be aware that this documentation section depends on the decision that `cookiecutter-plone-starter`is the way to move forward for an 'all-in-one' (backend and frontend) installer of Plone 6 from packages.
> 
> Indeed. I rarely get excited about software, but this way of doing things was so compelling that I tried to explain it to my wife over dinner last night.

ale-rt commented:
Sorry, misclicked

stevepiercy committed:
Add Cookiecutter to pre-requisites, and sort list.
Move Cookiecutter installation into pre-requisites and update to use user's Python.
Remove creation of a virtual environment, as that is handled by the cookiecutter.
Update cookiecutter command with correct feature branch.
Update console output.

stevepiercy committed:
Replace two options to configure Zope with a single sentence.

stevepiercy committed:
List mxdev as a pre-requisite and install in user's Python

stevepiercy committed:
Update mxdev files with installed defaults, with a single modification for collective.easyform.
Update instructions to use make instead of mxdev.

stevepiercy commented:
One more final round of reviews, and I think this can be merged. w00t!

ksuess commented on docs/install/install-from-packages.md:
````diff
@@ -0,0 +1,288 @@
+
+```shell
+cookiecutter https://github.com/collective/cookiecutter-plone-starter.git --checkout feature-4
+```
````
@jensens and me integrated mxdev in cookiecutter-plone-starter and tailored the `make` stuff. 
The checkout of the branch can be removed.

ksuess commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,505 @@
+	@echo "$(GREEN)==> Install Plone and create instance$(RESET)"
+	bin/pip install Plone -c https://dist.plone.org/release/$(PLONE_VERSION)/constraints.txt
+	bin/cookiecutter -f --no-input --config-file instance.yaml https://github.com/plone/cookiecutter-zope-instance
+	mkdir -p var/{filestorage,blobstorage,cache,log}
```
```suggestion
bin/pip:
	@echo "$(GREEN)==> Setup Virtual Env$(RESET)"
	python3 -m venv .
	bin/pip install -U "pip" "wheel" "cookiecutter" "mxdev"

instance/etc/zope.ini:	bin/pip
	@echo "$(GREEN)==> Install Plone and create instance$(RESET)"
	bin/cookiecutter -f --no-input --config-file instance.yaml https://github.com/plone/cookiecutter-zope-instance
	mkdir -p var/{filestorage,blobstorage,cache,log}

.PHONY: build-dev
build-dev: instance/etc/zope.ini ## pip install Plone packages
	@echo "$(GREEN)==> Setup Build$(RESET)"
	bin/mxdev -c mx.ini
	bin/pip install -r requirements-mxdev.txt
```

This has changed a little bit. 
We moved mxdev stuff to build-dev which replaces `pip install -r req…`
`instance/etc/zope.ini` cares for Zope **configuration**, but not for installing packages. Installing packages with versions, checkout, etc is the job of **mxdev** configuration

ksuess commented on docs/install/install-from-packages.md:
```diff
@@ -0,0 +1,288 @@
+
+-   Python 3.8, 3.9, or 3.10.
+-   Cookiecutter
+-   mxdev
```
This is handled by cookiecutter-plone-starter.

ksuess commented on docs/install/install-from-packages.md:
````diff
@@ -0,0 +1,288 @@
+
+```shell
+pip install --user --upgrade cookiecutter mxdev
+```
````
```suggestion
```shell
pip install --user --upgrade cookiecutter
```
```

mxdev is installed by cookiecutter-plone-starter.

ksuess commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,505 @@
+
+(manage-backend-installation-details-label)=
+
+## Backend installation details
```
This information but no instruction. Maybe move to the end of the chapter.

`instance/etc/zope.ini`is not a make target a developer needs to know if not interested in details. It is run by build-backend

ksuess committed:
mxdev is installed by cookiecutter-plone-starter.

ksuess committed:
mxdev is integrated in cookiecutter-plone-starter

ksuess committed:
Add mxdev to installation tools

ksuess committed:
Update 'Backend installation details' and move to additional info

ksuess committed:
appliying changes with  `make build-backend`

ksuess committed:
highlight code line

ksuess committed:
Update manage-add-ons-packages.md

To be discussed, but I thinks for this instructions a pragmatic narration code makes sense.

ksuess committed:
Move 'add add-on' enhancement to 'Common management tasks'

appliying changes is done by ``make build-backend`

ksuess committed:
Add link to 'commone management tasks'

ksuess committed:
simplify to cookiecutter-plone-starter make

ksuess committed:
Everything explained here in 'common tasks'

ksuess committed:
Update manage-add-ons-packages.md: apply changes, rebuild and restart backend

ksuess committed:
Add instructions on re-installing Plone add-ons.

ksuess reviewed (commented):
Moved to commits.

ksuess commented:
Thank you @stevepiercy for reviewing my work on 'Installation of Plone 6 for developers'. 
Please have a look and review once more, thakns.

ksuess commented:
please review, @jensens

jensens commented on docs/install/install-from-packages.md:
````diff
@@ -0,0 +1,279 @@
+Install or upgrade {term}`Cookiecutter` and {term}`mxdev` in your user's Python:
+
+```shell
+pip install --user --upgrade cookiecutter
````
One line above it is telling me to install mxdev, so:
```suggestion
pip install --user --upgrade cookiecutter mxdev
```

ksuess replied:
if i am not completly wrong mxdev is installed by cookiecutter-plone-starter. your commit…

jensens replied:
Ok, true, I should have noticed this myself. In this case my complaint is not completely wrong, because in line 52 there should be no mention of mxdev.

ksuess replied:
yes, thank you.

jensens commented on docs/install/install-from-packages.md:
```diff
@@ -0,0 +1,279 @@
+
+Now you can edit content or configure your Plone site.
+
+You can stop the site with {kbd}`ctrl-esc`.
```
Escape? I would go for `ctrl-c`
```suggestion
You can stop the site with {kbd}`ctrl-c`.
```

jensens commented on docs/install/install-from-packages.md:
````diff
@@ -0,0 +1,279 @@
+2022-09-18 23:14:31,893 INFO    [waitress:486][MainThread] Serving on http://127.0.0.1:8080
+```
+
+At this point, you could visit the Plone Classic UI interface at the URL http://localhost:8080.
+However it is strongly recommended to use the new Plone frontend, Volto, to gain many advantages, including easier content editing and development of components.
+Continue to the next section to do so.
````
I would put this in an info box or skip here at all. I fact we are telling what one could do, but better not do. This is for sure confusing.

stevepiercy replied:
Good point. Deleted. The Volto frontend is almost in parity with the Classic UI, and that's the only reason I put it there in the first place. We can address any lack of parity by fixing the bugs instead of mentioning it in the docs.

stevepiercy replied:
Deleted in 7c30797e40d105a901f5d6874b45bd9812da065c

jensens commented on docs/install/install-from-packages.md:
````diff
@@ -0,0 +1,279 @@
+cd my_project
+```
+
+Run `cookiecutter` to create a Plone project skeleton using the cookiecutter {term}`cookiecutter-plone-starter` with the following command.
````
If one is not familiar with cookicutter the following sentence is over-cookiecuttered:
```suggestion
Run `cookiecutter` to create a Plone project skeleton using the template {term}`cookiecutter-plone-starter` with the following command.
```

ksuess committed:
Nice but with a focus on develoment this is not needed here.

Sorry for the noise, I brought that in and suggest to remove it.

jensens reviewed (commented):
Overall almost ready for merge, I added some nit-picky comments.

ksuess committed:
Pin the version of an add-on

ksuess committed:
Add info on upgrade step execution for Plone core packages

ksuess committed:
run backend with make

ksuess committed:
update link to backend configuration

ksuess committed:
update link to backend configuration (follow up)

ksuess committed:
placeholder for frontend: add add-on, checkout add-on.

ksuess committed:
Apply suggestions from code review

Co-authored-by: Jens W. Klein <jk@kleinundpartner.at>

ksuess committed:
update todo for frontend add-ons

ksuess committed:
Merge branch 'installation-from-source-review' of github.com:plone/documentation into installation-from-source-review

ksuess committed:
Remove mxdev from installation instruction.

stevepiercy committed:
Remove TODO about pre-requisites, add GNU make install references.

stevepiercy committed:
Clean up make and cookiecutter terminology, anonymize console output

stevepiercy committed:
Remove distraction about visiting backend

stevepiercy committed:
Update console output from make start-backend

stevepiercy committed:
Remove pm2, unnecessary sentence.

stevepiercy committed:
Remove processes from heading

stevepiercy committed:
The install details in this section are limited to cookiecutter

stevepiercy committed:
Align label with heading

stevepiercy committed:
Update heading and label to reflect content.

stevepiercy committed:
Grammar fixes

stevepiercy committed:
This section is an overview of mxdev usage.

stevepiercy committed:
Add location of mxdev files

stevepiercy committed:
Revise introduction to Common management tasks section

stevepiercy committed:
Revise Add an add-on section, referring to default mxdev files, modifying them, and make narrative text more generic for any add-on

stevepiercy commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,400 @@
+
+In your web browser, and assuming you are currently logged in as `admin`, visit the URL http://localhost:8080/Plone/prefs_install_products_form.
+
+At the top of the page, you should see an upgrade information. Update your add-on.
```
I ran through this process, and there is no such information.
![Screen Shot 2022-09-24 at 3 18 17 AM](https://user-images.githubusercontent.com/102112/192092702-34bfddc5-f9e9-46d3-ad02-a31ee51d7231.png)

There is, however, outdated information referring to buildout.

```suggestion
```

ksuess replied:
Yes, difficult to document. If there is an upgrade step, depends on which package version was installed before and if there is an upgrade step for an upgrade from former version to new version.
So, maybe something like "If there is an upgrade step, please execute it." would be helpful.

If there is an upgrade step, then the add-on control panel http://localhost:8080/Plone/prefs_install_products_form would look like the following. I installed collective.easyform 3.1.0. Then upgraded to collective.easyform 4.1.0
![Plone_upgrade_step](https://user-images.githubusercontent.com/1144737/192202772-856698cc-83f7-495b-b39a-3b627b91ec1e.png)

stevepiercy replied:
Got it. I think this commit handles this difficult to explain scenario. Please let me know. 8651e8e3f302bef13ceac84def634449267c5acd

stevepiercy committed:
Update "Pin the version of an add-on" section

stevepiercy commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,400 @@
+
+In your web browser, and assuming you are currently logged in as `admin`, visit the URL http://localhost:8080/Plone/prefs_install_products_form.
+
+Then click the {guilabel}`Install` button to complete installation of `collective.easyform`.
```
Strangely, for this one, the upgrade button appeared! I clicked it and the console:

```
2022-09-24 03:27:55,830 INFO    [collective.easyform.upgrades:92][waitress-1] Removed easyform.migrate_all_forms registry record
2022-09-24 03:27:55,839 INFO    [GenericSetup.controlpanel:100][waitress-1] Control panel imported.
2022-09-24 03:27:55,840 INFO    [Products.GenericSetup.tool:1207][waitress-1] Profile collective.easyform:default upgraded to version '1014'.
```

ksuess replied:
Same here. It depends which version was installed and if there is an upgrade step from old version to new version. This is up to the developer of the package. Upgrade steps are needed for example if a new index is introduced and the catalog needs to be updated. See https://training.plone.org/5/mastering-plone/upgrade_steps.html?highlight=upgrade%20step

But you already found this info!

stevepiercy replied:
I'm going to resolve this conversation, as it seems to be a duplicate of https://github.com/plone/documentation/pull/1277#discussion_r979655667

stevepiercy committed:
Fix labels

stevepiercy committed:
Update "Check out an add-on" section

stevepiercy committed:
Update "Check out a Plone package" section

stevepiercy committed:
Update "Backend installation details" section

stevepiercy commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,425 @@
+Add info here or link to respective frontend chapter.
+localhost:8050/volto/addons/index.html#configuring-a-volto-project-to-use-an-addon
+This already exists above:
+For developing add-ons for the Plone frontend, Volto, see {doc}`volto/addons/index`.
```
@ksuess see the above two lines.

Is this todo needed here? I would prefer to remove it from the documentation, and create an issue.

ksuess replied:
Yes, please.

stevepiercy replied:
Created issue: https://github.com/plone/documentation/issues/1326

Removed todo in 748cc7093c57647eebb4a0544b51fa012f4e2957

stevepiercy reviewed (commented):
@ksuess I have a few questions for you.

stevepiercy commented:
I made another round of changes. Review and feedback welcome. @ksuess I have a few specific questions for you in the review. Thank you!

stevepiercy committed:
Remove todo about frontend configuration, and create issue https://github.com/plone/documentation/issues/1326

stevepiercy committed:
Revise instructions to accommodate either an upgrade or install of a package or add-on.

stevepiercy commented:
@ksuess thank you! If I understood correctly what you explained, I pushed a couple of commits to resolve the last bits. Please review. Thank you!

ksuess commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,407 @@
+
+### Build and start your instance
+
+Build and run the Plone backend.
```
Sorry, just one more.
A build is necessary on changes like mentioned in sections above (another add-on needed, etc) So maybe one sentence like "You made changes, cause you need another add-on, you forced to override a Plone core package, etc.. Then a build and restart is needed." would illustrate the need for a build and restart .

stevepiercy replied:
Good point. Added explanation for why it is needed in ce35ee48ed10078f3628ee91e2f9ab5e04d96a69

ksuess reviewed (changes requested):
Sorry, just one more.

ksuess commented on docs/install/manage-add-ons-packages.md:
```diff
@@ -0,0 +1,407 @@
+
+## Backend installation details
+
+The `Makefile` at the root of you project invokes commands in `backend/Makefile`.
```
```suggestion
The `Makefile` at the root of your project invokes commands in `backend/Makefile`.
```

stevepiercy replied:
Fixed in d24df6a800972d0e83c58b89372a89034c73e190

stevepiercy committed:
Clarify when it is necessary to build and restart the backend

stevepiercy committed:
Fix typo

stevepiercy commented:
@ksuess one more time, please!

ksuess reviewed (approved):
Great! Love it, @stevepiercy
